### PR TITLE
Better JSDoc types

### LINKED ISSUE #6488: JSDoc types for addons
### Why this would be helpful

The new schema for addon.json is very helpful as it allows me to be much faster writing my addon.json files with autocomplete. This is, however, not currently available for JS addons.

### My suggestion

My suggestion is to create JSDoc types that can be utilized like so:

```js
import "../../types/addon.js";

/** @type {AddonFn} */
export default async function ({ addon, console, msg }) {
  // ...
}
```

and provide valuable autocomplete.

### Possible alternatives

_No response_

### Additional context

I'm working on this and am actually finished with everything but `addon.tab`, which happens to have the most convoluted function definitions...

https://github.com/ScratchAddons/ScratchAddons/assets/130385691/d2ab753c-64c6-4941-8bec-280f6de8079c

Apologies for the low quality video, not sure what happened there.

### PR DESCRIPTION
**Pull request by mybearworld**
Resolves #6488
Closes ScratchAddons/website-v2#337
Part of #6721

### Changes

Improved the JSDoc types and grammar, and also added the JSDoc types to every addon.

### Reason for changes

Better DX.

### Tests

Tested in VSCode - browser testing isn't really applicable here

### CODE REVIEW

**Comment by mybearworld** [comment]
Related: ScratchAddons/website-v2#337

**Comment by WorldLanguages** [comment]
Sorry for not checking this before, I will try to find some time next month.

**Comment by mybearworld** [comment]
All that's necessary to get proper type defs now is:
```js
/** @param {AddonAPI} */
```

**Comment by mybearworld** [comment]
199 files changed...

Actually, on that note, if this gets merged, the majority of addons would have this as their latest commit. This might cause some issues?

A couple of folders, where most of them have the commit message of "Revert \"Put type `AddonAPI` in global scope\"".

**Comment by BroJac5246** [comment]
> Actually, on that note, if this gets merged, the majority of addons would have this as their latest commit. This might cause some issues?

It has happened before and nothing bad has come of it so we're probably fine.

**Comment by WorldLanguages** [comment]
> Actually, on that note, if this gets merged, the majority of addons would have this as their latest commit. This might cause some issues?

To be honest, that is one of the reasons I sometimes lack the energy to review PRs like this, but it's not really an actual problem :P

**Comment by mybearworld** [comment]
> To be honest, that is one of the reasons I sometimes lack the energy to review PRs like this

Sorry :P Unfortunately, I don't really think there's another way to apply the types to all addons.

**Comment by WorldLanguages** [comment]
Indeed, there's no other way to define the argument types, as far as we know. If `addon` was a global object, it would be easier.

**Comment by Samq64** [comment]
> The conflicts are coming from #7308. Should I resolve them, or wait for the lines to be removed again?

Just wait. The whole PR will just be reverted.

**Comment by mybearworld** [comment]
Cool, the conflicts are gone with #7325

**Comment by mybearworld** [comment]
Fixed the merge conflicts introduced with #7376

**Comment by mybearworld** [comment]
Fixed the merge conflicts introduced with #7394

**Comment by mybearworld** [comment]
Oh no
image

**Comment by DNin01** [comment]
> Oh no
> 
> image

Can you resolve them with Git (or GitHub Desktop or VS Code)?

**Comment by mybearworld** [comment]
> Can you resolve them with Git (or GitHub Desktop or VS Code)?

Oh yeah, I wanted to do that, and then forgot :P Looks like it was because of a deleted file. I've resolved them

**Comment by mybearworld** [comment]
Should be up to date now

**Comment by WorldLanguages** [comment]
Maybe I can take a look for v1.41, I don't have enough time to do it before that

**Comment by Samq64** [comment]
> Indeed, there's no other way to define the argument types, as far as we know. If `addon` was a global object, it would be easier.

Does `__addon` now being available globally help anything?

**Comment by mybearworld** [comment]
> Does `__addon` now being available globally help anything?

Addons still get in their own `addon` (and `msg` and `console`) as function parameters, so I don't think so.

**Review by ajskateboarder — APPROVED** [review]

**Review by Supernoodles99 — APPROVED** [review]
engineer gaming

File: addons/redirect-mobile-forums/userscript.js
Comment (Samq64) [inline addons/redirect-mobile-forums/userscript.js:1]: This file can be deleted since the addon was moved to v2.

**Pull request closed without merging**